web/src/components/onboarding/OnboardingTaxId.tsx

```diff
@@ -100,65 +125,136 @@ export const OnboardingTaxId = ({
     props.onValidation && props.onValidation(fieldName, Object.values(errors(errorMap)).includes(true));
   };
 
-  const error = fieldStates[fieldName].invalid;
+  const setSplitTaxId = (taxId: string) => {
+    setTaxIdValue(taxId.trim().slice(0, 9));
+    setLocationValue(taxId.trim().slice(9, 12));
+  };
 
-  if (initialType.current === "FULL") {
+  const toggleTaxIdDisplay = async () => {
+    if (!state.profileData.taxId) {
+      return;
+    }
+    if (!state.profileData.taxId.includes(maskingCharacter) && !displayTaxId) {
+      if (initialType.current == "SPLIT") {
+        setSplitTaxId(state.profileData.taxId);
+      }
+      setDisplayTaxId(true);
+    } else if (
+      state.profileData.taxId.includes(maskingCharacter) &&
+      !displayTaxId &&
+      state.profileData.encryptedTaxId
+    ) {
+      const decryptedTaxId = await decryptTaxId({ encryptedTaxId: state.profileData.encryptedTaxId });
+      setProfileData({ ...state.profileData, taxId: decryptedTaxId });
+      if (initialType.current == "SPLIT") {
+        setSplitTaxId(decryptedTaxId);
+      }
+      setDisplayTaxId(true);
+    } else {
+      setDisplayTaxId(false);
+    }
+  };
```

Comment:
I agree, I will figure out a way to make it more readable tomorrow.

Reply:
consider our refactoring code smell "Repeated Switches" and its associated solution "replace conditional with polymorphism"

At least to me, looking at the return values of this component, it seems to be begging for two components for split/full instead of having one thing handle all these split/full conditional logic